torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -402,6 +429,58 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
   return futureResponseMessage;
 }
 
+void TensorPipeAgent::pollTimeoutRpcs() {
+  while (rpcAgentRunning_.load()) {
+    std::unique_lock<std::mutex> lock(timeoutMapMutex_);
+
+    steady_clock_time_point earliestTimeout =
+        std::chrono::steady_clock::now() + kLargeTimeDuration;
+
+    // We sleep until the earliest expiring RPC in the timeoutMap_. We must
+    // also ensure that we sleep while the map is empty, and we exit sleeping
+    // if the RPC Agent has been shutdown.
+    for (;;) {
+      if (!rpcAgentRunning_.load()) {
+        return;
+      }
+      if (std::chrono::steady_clock::now() >= earliestTimeout) {
+        break;
+      }
+      if (!timeoutMap_.empty()) {
+        earliestTimeout = timeoutMap_.begin()->first;
+      }
+      timeoutThreadCV_.wait_until(lock, earliestTimeout);
+    }
+
+    // Move all these futures to a separate vector so we can process them
+    // outside the lock.
+    std::vector<std::shared_ptr<FutureMessage>> timedOutFutures =
+        std::move(timeoutMap_.begin()->second);
+    // We can safely remove this key from the timeoutMap_ since all these
+    // futures will be processed.
+    timeoutMap_.erase(timeoutMap_.begin());
+
+    lock.unlock();
+
+    // Set an error on futures added to the timedOutFutures vector. We do this
+    // outside the lock to prevent potential lock-order-inversions by callbacks
+    // triggered by the serError call.
+    for (const auto& future : timedOutFutures) {
+      std::string errorMsg = c10::str(
+          "RPC ran for more than set timeout and will now be marked with an error");
+      // Using setErrorIfNeeded so completed futures are ignored.
+      future->setErrorIfNeeded(errorMsg);
+      if (future->hasError()) {
+        // We only want to decrease the active call count if the future has an
+        // error. If it was successful, the counter was already updated in the
+        // send() function above.
+        --clientActiveCalls_;
+      }
+    }
+    timedOutFutures.clear();
```

Comment:
Actually, no need to explicitly clear it: the vector will be automatically destroyed as soon as we leave the scope in the next line.